legate/cpp/legate_kvikio.cpp

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright (c) 2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cassert>
+#include <cstddef>
+#include <cstdint>
+#include <iostream>
+#include <sstream>
+#include <utility>
+
+#include <legate_library.h>
+#include <legate_kvikio.hpp>
+
+#include <kvikio/file_handle.hpp>
+
+namespace legate_kvikio {
+
+struct elem_size_fn {
+  template <legate::LegateTypeCode DTYPE>
+  size_t operator()()
+  {
+    using dtype = typename legate::LegateTypeOf<DTYPE>::type;
+    return sizeof(dtype);
+  }
+};
+
+size_t sizeof_legate_type_code(legate::LegateTypeCode code)
+{
+  return legate::type_dispatch(code, elem_size_fn{});
+}
+
+class WriteTask : public Task<WriteTask, TaskOpCode::OP_WRITE> {
+ public:
+  static void cpu_variant(legate::TaskContext& context)
+  {
+    std::string path     = context.scalars()[0].value<std::string>();
+    legate::Store& store = context.inputs()[0];
+    auto shape           = store.shape<1>();
+    auto acc             = store.read_accessor<char, 1>();
+    size_t strides[1];
+    const char* data = acc.ptr(shape, strides);
```

Comment:
@manopapad, can I assume that `data` is contiguous? 
If not, can I create two task variants - one to handle contiguous stores and one for non-contiguous stores?

Reply:
> can I assume that data is contiguous?

You will certainly want to check this, since the API allows the user to pass in an arbitrary array (if the API were to create its own output array, then you could guarantee it internally). You could run into trouble if the user did something like:

```
x = np.zeros((4,5))
read_into(x[:,3])
```

where the user passes in a column, which is technically 1d, but that slice is not contiguous. Even in this case, you can force Legion to give you a "compacted" view of the slice, by declaring that the task needs an "exact" instance, similar to how it's done here https://github.com/nv-legate/cunumeric/blob/branch-23.05/src/cunumeric/mapper.cc#L111.

> If not, can I create two task variants - one to handle contiguous stores and one for non-contiguous stores?

Currently Legate's variant registration system is specialized for making 3 specific variants (CPU, OpenMP, GPU), so we'd  need some work to extend it beyond that. The simpler alternative, if you want to support non-contiguous stores, would be to take cases inside the task body as written.